reactioncommerce/reaction
# fix: 4741 catalog variant inventory flags always false

Resolves #4741  
Impact: **major**  
Type: **bugfix**

## Issue

`isSoldOut` and `isLowQuantity` for variants and options in the catalog were always `false`, and not reflective of their actual status.

## Solution

Fix the transform that creates the catalog variants by processing the correct inventory flags just like it's done for the top-level product

## Breaking changes

Added a new, final param to `xformVariant` with the processed inventory flags

## Testing

Try to set the quantities and warn at thresholds to various variants and options and check the DB or GQL for the expected results.

For example:
- If you set a variant quantity to 0, it should be sold out if it has no options
- If you set quantities of the options of a  variant  to 0, each option should be sold out as well as the top-level variant

**Play around with different combinations of sold out, low inventory and back ordered**

**Below are some variations**

---

1. With a product that only has top-level variants
2. Set the quantity to 0 and publish
3. Query in GQL or Mongo for that catalog item variant
4. See that `isSoldOut` is `true`

---

1. With a product that only has top level variants
2. Set the low quantity threshold to 1000 and publish
3. Query in GQL or Mongo for that catalog item variant
4. See that `isLowQuantity` is `true`

---

1. With a product that has variant options
2. Set each of the options quantity to 0 and publish
3. Query in GQL or Mongo for that catalog item variant that contains those options
4. See that `isSoldOut` is `true`
5. See that `isSoldOut` is `true` for the option itself

---

1. With a product that has variant options
2. Set the low quantity threshold to 1000 on the top-level variant and publish
3. Query in GQL or Mongo for that catalog item variant that contains those options
4. See that `isLowQuantity` is `true` for the option itself

---

1. Checkout with a product that has a quantity of 1
1. Approving or canceling the order in the order panel should update inventory status accordingly

---

1. Checkout with a product that has a quantity of 11 and a `warn at` at 10
1. Approving or canceling the order in the order panel should update inventory status accordingly

---


**Test the Migrations**

1. Start reaction with this branch
2. `mongorestore --drop` the `reaction-2.0.0-rc.1` mongo dump
3. Run a graph QL query for a product that has legacy variant inventory flags, or check the catalog collection.

```
{
  catalogItemProduct(slugOrId: "converse-all-stars-black") {
    product {
      variants {
        title
        isSoldOut
        isLowQuantity
        inventoryPolicy
        inventoryManagement
        options {
          title
          isSoldOut
          isLowQuantity
        }
      }
    }
  }
}
```
4. See the variant and variant options `isSoldOut` are `false`. This is incorrect. 
5. Stop and start reaction again
6. Wait for migration 42 to run
7. Re-run the query and see the variant and variant options `isSoldOut` are now `true`
